Alerting: Add pagination and improved search for notification policies

soniaAguilarPeiron opened the pull request:
**What is this feature?**

This PR adds several improvements to the policies view.

So far, if we have a big number of policies, the UI becomes very slow in this page. The interaction in this page becomes pretty irresponsive.

With this PR we introduce several things that fix this problem and also improve the user experience when searching.

- For each policy in the tree that exceed X number of policies, we add some pagination for its children with `{numberofpoliciesHidden} additional policies` button to expand the list of the branch that exceeds this X number
- When using the filters, the UI now shows a feedback of the results indicating if there are no matches or the number of policies that match the filters.
- When using filters, UI only show the nodes that belong to a matching path.


**Why do we need this feature?**

UI should support big number of policies.

**Who is this feature for?**

All users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/73038

**Special notes for your reviewer:**


https://github.com/grafana/grafana/assets/33540275/dd8b081a-953c-4982-85eb-e831b847007c


**About the filter results:**


https://github.com/grafana/grafana/assets/33540275/32af5b79-791d-4124-9b4d-55a2fa98b84f



Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

soniaAguilarPeiron committed:
Add show more pagination and search feedback to policies view

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/policy-view-improvements

gillesdemey commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -227,11 +230,22 @@ const AmRoutes = () => {
-                      onChangeMatchers={setLabelMatchersFilter}
-                      onChangeReceiver={setContactPointFilter}
-                    />
+                    <Stack direction="row" gap={1}>
```
Perhaps we want to move this logic to the `NotificationPoliciesFilter`, passing a number of matches as a property;

1. It already has some code to determine if we have any filters (though uses query params)
1. We can then wrap the `<Button>` and the results with a `<Stack alignItems="center">`
2. There's an odd delay right now between the button disappearing and the number of matches which would be eliminated (probably because query strings are used to determine rendering of the clear button)
3. keeps the logic here a bit cleaner since it doesn't need to be concerned about rendering anything related to filters

gillesdemey commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -141,6 +144,10 @@ const Policy = (props: PolicyComponentProps) => {
     errors.push(error);
   });
 
+  const POLICIES_PER_PAGE = 20;
```
nit: maybe hoist this variable to the top of the page :)

gillesdemey commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -99,6 +99,9 @@ const Policy = (props: PolicyComponentProps) => {
   const hasMatchers = Boolean(matchers && matchers.length);
   const hasFocus = routesMatchingFilters.some((route) => route.id === currentRoute.id);
+  const childrenNodesIds = currentRoute.routes?.map((route) => route.id) ?? [];
+  const someChildrenMatchSearch =
```
Since this is a draft; I think we should not highlight parent policies with matching children in the same way we do the _actual_ matches. The screenshot below is I think a good indicator for why that would be confusing :)

The parent policy isn't using the `grafana-default-email` contact point but is still highlighted, I find that a bit confusing :D

<img width="1075" alt="image" src="https://github.com/grafana/grafana/assets/868844/12d0d208-c344-4c0e-8f9c-52b405365fd5">

soniaAguilarPeiron replied:
maybe we could use another color for the parent?

gillesdemey replied:
Let's table this for now and focus on the pagination and indicators for how many matchers are found :)

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/policy-view-improvements

soniaAguilarPeiron committed:
move results info to the policies filter component

soniaAguilarPeiron committed:
Hide not matching nodes in policy tree
Co-authored-by: Gilles De Mey <gilles.de.mey@gmail.com>
Co-authored-by: Konrad Lalik <konradlalik@gmail.com>

soniaAguilarPeiron committed:
Fix test

gillesdemey committed:
remove unnecessary styles and flex wrapper

gillesdemey committed:
move hasFilters to RoutesMatchingFilters

gillesdemey committed:
load more policies styles

I don't think it makes sense to allow showing fewer policies – why did the user click it in the first place if they weren't interested?

gillesdemey committed:
remove unused code

gillesdemey committed:
button flex styles

gillesdemey committed:
styles

soniaAguilarPeiron committed:
Add tests for the new map in findRoutesMatchingFilters function

soniaAguilarPeiron committed:
Merge remote-tracking branch 'origin/main' into alerting/policy-view-improvements

soniaAguilarPeiron commented:
/deploy-to-hg

gillesdemey committed:
tiny text update

gillesdemey committed:
beterrer

gillesdemey committed:
add assertions for filtersApplied

gillesdemey committed:
beterrer

soniaAguilarPeiron commented:
/deploy-to-hg

konrad147 commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -274,41 +284,90 @@ type RouteFilters = {
+
+export interface RoutesMatchingFilters {
+  filtersApplied: boolean;
+  matchedRoutes: RouteWithID[];
```
What additional information does it provide over `matchedRoutesWithPath`?

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -102,7 +110,15 @@ const Policy = (props: PolicyComponentProps) => {
+  const { filtersApplied, matchedRoutes, matchedRoutesWithPath } = routesMatchingFilters;
+
+  // check if this route matches the filters
+  const hasFocus = filtersApplied && matchedRoutes.some((route) => route.id === currentRoute.id);
+
+  // check if this route belongs to a path that matches the filters
+  const routesPath = Array.from(matchedRoutesWithPath.values()).flat();
+  const belongsToMatchPath = routesPath.some((route: RouteWithID) => route.id === currentRoute.id);
 
```
I'm a bit concerned that this code can cause performance issues. Each `Policy` component needs to loop over `matchedRoutes` and `matchedRoutesWithPath` . Performance-wise it would be better to, for example build a `Set` object of matching routes Ids and pass it to each `Policy` so it's not recalculated over and over again

soniaAguilarPeiron replied:
I tested with 5000 policies and it was pretty fast. Maybe there is a use case I'm missing?

konrad147 replied:
I think that if each policy had multiple matchers and if the filter would match most of policies you have

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -162,12 +191,33 @@ const Policy = (props: PolicyComponentProps) => {
+    ? childPolicies.slice(0, visibleChildPolicies) // no filters, show all
+    : hasSomeChildPoliciesBelongingToMatchPath // if we have filters, show all that match the filters
+      ? childPoliciesBelongingToMatchPath.slice(0, visibleChildPolicies)
+      : [];
```
What is the reason for the second condition?

Is it different from:
```js
const childPoliciesToRender = !filtersApplied
    ? childPolicies.slice(0, visibleChildPolicies) // no filters, show all
    :  childPoliciesBelongingToMatchPath.slice(0, visibleChildPolicies)
```

Also, I'd consider switching the condition to be positive, so instead `!filtersApplied` just `filtersApplied`

konrad147 commented on public/app/features/alerting/unified/components/notification-policies/Policy.tsx:
```diff
@@ -162,12 +191,33 @@ const Policy = (props: PolicyComponentProps) => {
   const isImmutablePolicy = isDefaultPolicy || isAutogeneratedPolicyRoot;
   // TODO dead branch detection, warnings for all sort of configs that won't work or will never be activated
+
+  const childPoliciesBelongingToMatchPath = childPolicies.filter((child) =>
```
What if we split filtering and pagination into two steps? 
First step would be to get all filtered child policies or all when no filters applied
The second step would be only about pagination on the result of the above

gillesdemey commented:
Resolved @konrad147 's questions during 🍐-programming session

soniaAguilarPeiron committed:
Refactor
Co-authored-by: Gilles De Mey <gilles.de.mey@gmail.com>
Co-authored-by: Konrad Lalik <konradlalik@gmail.com>

konrad147 reviewed (approved):
LGTM!
